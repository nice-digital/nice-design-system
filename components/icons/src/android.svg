<?xml version="1.0" encoding="utf-8"?>
<svg width="512" height="512" viewBox="0 0 512 512" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M199.704 127.577C202.575 127.577 205.042 126.545 207.106 124.481C209.17 122.417 210.202 119.949 210.202 117.077C210.202 114.205 209.17 111.737 207.106 109.673C205.042 107.609 202.575 106.577 199.704 106.577C196.833 106.577 194.41 107.609 192.436 109.673C190.462 111.737 189.475 114.205 189.475 117.077C189.475 119.949 190.462 122.417 192.436 124.481C194.41 126.545 196.833 127.577 199.704 127.577ZM313.296 127.577C316.167 127.577 318.59 126.545 320.564 124.481C322.538 122.417 323.525 119.949 323.525 117.077C323.525 114.205 322.538 111.737 320.564 109.673C318.59 107.609 316.167 106.577 313.296 106.577C310.425 106.577 307.957 107.609 305.894 109.673C303.83 111.737 302.798 114.205 302.798 117.077C302.798 119.949 303.83 122.417 305.894 124.481C307.957 126.545 310.425 127.577 313.296 127.577ZM94.7251 177.115C102.262 177.115 108.722 179.808 114.106 185.192C119.489 190.577 122.181 197.038 122.181 204.577V320.346C122.181 328.064 119.534 334.615 114.24 340C108.947 345.385 102.442 348.077 94.7251 348.077C87.0088 348.077 80.4588 345.385 75.0753 340C69.6918 334.615 67 328.064 67 320.346V204.577C67 197.038 69.6918 190.577 75.0753 185.192C80.4588 179.808 87.0088 177.115 94.7251 177.115ZM380.052 182.231V361.538C380.052 369.795 377.181 376.795 371.438 382.538C365.696 388.282 358.787 391.154 350.712 391.154H330.523V452.269C330.523 459.987 327.832 466.538 322.448 471.923C317.065 477.308 310.515 480 302.798 480C295.082 480 288.532 477.308 283.148 471.923C277.765 466.538 275.073 459.987 275.073 452.269V391.154H237.927V452.269C237.927 459.987 235.235 466.538 229.852 471.923C224.468 477.308 217.918 480 210.202 480C202.665 480 196.205 477.308 190.821 471.923C185.437 466.538 182.746 459.987 182.746 452.269L182.477 391.154H162.558C154.303 391.154 147.304 388.282 141.562 382.538C135.819 376.795 132.948 369.795 132.948 361.538V182.231H380.052ZM317.603 73.1923C336.804 83.0641 352.147 96.8397 363.632 114.519C375.117 132.199 380.859 151.538 380.859 172.538H131.871C131.871 151.538 137.614 132.199 149.099 114.519C160.584 96.8397 176.016 83.0641 195.397 73.1923L176.286 37.9231C175.029 35.5897 175.478 33.7949 177.631 32.5385C179.964 31.4615 181.759 32 183.015 34.1538L202.396 69.6923C219.443 62.1538 237.478 58.3846 256.5 58.3846C275.522 58.3846 293.557 62.1538 310.604 69.6923L329.985 34.1538C331.241 32 333.036 31.4615 335.369 32.5385C337.522 33.7949 337.971 35.5897 336.714 37.9231L317.603 73.1923ZM446 204.577V320.346C446 328.064 443.308 334.615 437.925 340C432.541 345.385 425.991 348.077 418.275 348.077C410.738 348.077 404.278 345.385 398.894 340C393.511 334.615 390.819 328.064 390.819 320.346V204.577C390.819 196.859 393.511 190.353 398.894 185.058C404.278 179.763 410.738 177.115 418.275 177.115C425.991 177.115 432.541 179.763 437.925 185.058C443.308 190.353 446 196.859 446 204.577Z" fill="black"/>
</svg>
