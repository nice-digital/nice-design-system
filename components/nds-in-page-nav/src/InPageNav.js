import React from "react";
import PropTypes from "prop-types";

import "./../scss/in-page-nav.scss";

export const InPageNav = props => {
	return <p>TODO</p>;
};

InPageNav.propTypes = {};

InPageNav.defaultProps = {};

export default InPageNav;
