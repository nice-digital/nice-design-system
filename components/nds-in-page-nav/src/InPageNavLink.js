/* eslint-disable indent */
import React from "react";
import PropTypes from "prop-types";

export const InPageNavLink = ({
	link: { subLinks, href, title },
	activeHeadingId
}) => {
	const targetId = href.replace("#", "");

	const hasActiveSubLink = subLinks.some(
		(subLink) => subLink.href.replace("#", "") === activeHeadingId
	);

	return (
		<li className="in-page-nav__item">
			<a
				id={`inpagenav-${targetId}`}
				href={href}
				aria-current={
					targetId === activeHeadingId
						? "location"
						: hasActiveSubLink
						? "true"
						: null
				}
			>
				{title}
			</a>
			{subLinks && subLinks.length > 0 && (
				<ol
					className="in-page-nav__list"
					aria-labelledby={`inpagenav-${targetId}`}
				>
					{subLinks.map((subLink) => (
						<InPageNavLink
							key={subLink.href}
							link={subLink}
							activeHeadingId={activeHeadingId}
						/>
					))}
				</ol>
			)}
		</li>
	);
};

InPageNavLink.propTypes = {
	link: PropTypes.shape({
		title: PropTypes.string.isRequired,
		href: PropTypes.string.isRequired,
		subLinks: PropTypes.arrayOf(
			PropTypes.shape({
				title: PropTypes.string.isRequired,
				href: PropTypes.string.isRequired
			})
		)
	}).isRequired,
	activeHeadingId: PropTypes.string
};
