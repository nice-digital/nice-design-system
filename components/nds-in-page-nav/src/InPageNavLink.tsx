import React from "react";

export interface InPageNavLinkProps {
	link: {
		title: string;
		href: string;
		subLinks?: {
			title: string;
			href: string;
		}[];
	};
	activeHeadingId?: string | null;
}

export const InPageNavLink = ({
	link: { subLinks, href, title },
	activeHeadingId
}: InPageNavLinkProps) => {
	const targetId = href.replace("#", "");

	const hasActiveSubLink = subLinks?.some(
		(subLink) => subLink.href.replace("#", "") === activeHeadingId
	);

	return (
		<li className="in-page-nav__item">
			<a
				id={`inpagenav-${targetId}`}
				href={href}
				aria-current={
					targetId === activeHeadingId
						? "location"
						: hasActiveSubLink
						? "true"
						: undefined
				}
			>
				{title}
			</a>
			{subLinks && subLinks.length > 0 && (
				<ol
					className="in-page-nav__list"
					aria-labelledby={`inpagenav-${targetId}`}
				>
					{subLinks.map((subLink) => (
						<InPageNavLink
							key={subLink.href}
							link={subLink}
							activeHeadingId={activeHeadingId}
						/>
					))}
				</ol>
			)}
		</li>
	);
};
