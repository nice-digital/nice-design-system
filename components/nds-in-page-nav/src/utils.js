import slugify from "slugify";

const numberPrefixRegex = /^\s*\d+\s*/;

/**
 * Generates a unique id for the given heading HTML element and sets its id attribute
 * @param {HTMLHeadingElement} headingEl
 * @returns The id of the heading
 */
export const generateHeadingId = headingEl => {
	if (headingEl.id) return headingEl.id;

	// Remove numbers from start of string before slugifying to make sure we have a valid id
	const text = (headingEl.textContent || headingEl.innerText).replace(
		numberPrefixRegex,
		""
	);

	const slug = slugify(text, {
		lower: true,
		strict: true
	});

	if (!document.getElementById(slug)) {
		headingEl.id = slug;
		return headingEl.id;
	}

	// Increment the integer suffix to make a unique id
	const slugPrefixRegex = new RegExp(`^${slug}(-[0-9]+)?`, "i"),
		isSlugPrefixMatch = el => (el.id ? el.id.match(slugPrefixRegex) : false),
		matchingElements = Array.prototype.filter.call(
			document.querySelectorAll("[id]"),
			isSlugPrefixMatch
		),
		max = matchingElements.length;

	headingEl.id = `${slug}-${max + 1}`;
	return headingEl.id;
};

/**
 * Recursively builds a nested tree of links, from a flat list of heading elements.
 * @example
 * 	h2, h3, h3, h4, h2
 * Becomes:
 * 	h2
 * 	 h3
 * 	 h3
 * 	   h4
 * 	h2
 * @param {Array<HTMLHeadingElement>} headingElements The flat list of heading elements from which to build the tree
 * @param {Number | undefined} currentLevel The current heading level in the tree
 * @returns A tree of links with sublinks
 */
export const buildLinkTree = (headingElements, currentLevel = undefined) => {
	if (!headingElements || headingElements.length === 0) return [];

	const linkTree = [];

	if (!currentLevel)
		currentLevel = parseInt(headingElements[0].tagName.substring(1));

	for (var i = 0; i < headingElements.length; i++) {
		const heading = headingElements[i],
			headingLevel = parseInt(heading.tagName.substring(1));

		if (headingLevel === currentLevel) {
			linkTree.push({
				title: heading.textContent,
				href: "#" + generateHeadingId(heading),
				// Recursively build up sub links
				subLinks: buildLinkTree(headingElements.slice(i + 1), currentLevel + 1)
			});
		} else if (headingLevel < currentLevel) return linkTree;
	}

	return linkTree;
};

const flattenTree = linkTree => {
	return linkTree.reduce((acc, curr) => {
		return acc.concat(curr).concat(flattenTree(curr.subLinks));
	}, []);
};

export const getActiveHeadingId = (linkTree, scrollTolerance) => {
	const headingsAboveViewport = flattenTree(linkTree)
		.map(({ href }) => {
			const heading = document.querySelector(href);
			return { heading, y: heading.getBoundingClientRect().top };
		})
		.filter(({ y }) => y - scrollTolerance <= 0);

	if (headingsAboveViewport.length === 0) return null;

	return headingsAboveViewport.reduce((prev, current) => {
		return prev.y > current.y ? prev : current;
	}).heading.id;
};
