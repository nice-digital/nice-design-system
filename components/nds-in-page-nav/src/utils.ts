import { string } from "prop-types";
import slugify from "slugify";

const numberPrefixRegex = /^\s*\d+\s*/;

/**
 * Generates a unique id for the given heading HTML element and sets its id attribute
 * @param {HTMLHeadingElement} headingEl
 * @returns The id of the heading
 */
export const generateHeadingId = (headingEl: HTMLHeadingElement): string => {
	if (headingEl.id) return headingEl.id;

	// Remove numbers from start of string before slugifying to make sure we have a valid id
	const text = (headingEl.textContent || headingEl.innerText).replace(
		numberPrefixRegex,
		""
	);

	const slug = slugify(text, {
		lower: true,
		strict: true
	});

	if (!document.getElementById(slug)) {
		headingEl.id = slug;
		return headingEl.id;
	}

	// Increment the integer suffix to make a unique id
	const slugPrefixRegex = new RegExp(`^${slug}(-[0-9]+)?`, "i"),
		isSlugPrefixMatch = (el: HTMLElement) =>
			el.id ? el.id.match(slugPrefixRegex) : false,
		matchingElements = Array.prototype.filter.call(
			document.querySelectorAll("[id]"),
			isSlugPrefixMatch
		),
		max = matchingElements.length;

	headingEl.id = `${slug}-${max + 1}`;
	return headingEl.id;
};

export interface LinkTreeItem {
	title: string;
	href: string;
	subLinks: LinkTreeItem[];
}

/**
 * Recursively builds a nested tree of links, from a flat list of heading elements.
 * @example
 * 	h2, h3, h3, h4, h2
 * Becomes:
 * 	h2
 * 	 h3
 * 	 h3
 * 	   h4
 * 	h2
 * @param {Array<HTMLHeadingElement>} headingElements The flat list of heading elements from which to build the tree
 * @param {Number | undefined} currentLevel The current heading level in the tree
 * @returns A tree of links with sublinks
 */
export const buildLinkTree = (
	headingElements: HTMLHeadingElement[],
	currentLevel: undefined | number = undefined
): LinkTreeItem[] => {
	if (!headingElements || headingElements.length === 0) return [];

	const linkTree = [];

	if (!currentLevel)
		currentLevel = parseInt(headingElements[0].tagName.substring(1));

	for (var i = 0; i < headingElements.length; i++) {
		const heading = headingElements[i],
			headingLevel = parseInt(heading.tagName.substring(1));

		if (headingLevel === currentLevel) {
			linkTree.push({
				title: heading.textContent as string,
				href: "#" + generateHeadingId(heading),
				// Recursively build up sub links
				subLinks: buildLinkTree(headingElements.slice(i + 1), currentLevel + 1)
			});
		} else if (currentLevel && headingLevel < currentLevel) return linkTree;
	}

	return linkTree;
};

const flattenTree = (linkTree: LinkTreeItem[]): LinkTreeItem[] => {
	return linkTree.reduce((acc, curr) => {
		return acc.concat(curr).concat(flattenTree(curr.subLinks));
	}, [] as LinkTreeItem[]);
};

export const getActiveHeadingId = (
	linkTree: LinkTreeItem[],
	scrollTolerance: number
): string | null => {
	const headingsAboveViewport = flattenTree(linkTree)
		.map(({ href }) => {
			const heading = document.querySelector(href);
			return { heading, y: heading?.getBoundingClientRect().top || 0 };
		})
		.filter(({ y }) => y - scrollTolerance <= 0);

	if (headingsAboveViewport.length === 0) return null;

	const reducedHeadings = headingsAboveViewport.reduce((prev, current) => {
		return prev.y > current.y ? prev : current;
	});

	return reducedHeadings?.heading?.id || null;
};
