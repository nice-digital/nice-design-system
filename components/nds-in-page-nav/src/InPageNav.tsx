/* eslint-disable no-mixed-spaces-and-tabs */
/* eslint-disable indent */
import React, { useEffect, useState } from "react";
import throttle from "lodash.throttle";

import { buildLinkTree, getActiveHeadingId, type LinkTreeItem } from "./utils";
import { InPageNavLink } from "./InPageNavLink";

import "./../scss/in-page-nav.scss";

export interface InPageNavProps {
	[prop: string]: unknown;
	className?: string;
	headingsContainerSelector?: string;
	headingsSelector?: string;
	headingsExcludeSelector?: string;
	headingsExcludeContainer?: string;
	scrollTolerance?: number;
}

export const InPageNav = ({
	className,
	headingsContainerSelector = "main",
	headingsSelector = "h2, h3",
	headingsExcludeSelector = "", // Exclude any matched elements
	headingsExcludeContainer = "", // Exclude all headings within this container
	scrollTolerance = 50, // In pixels
	...rest
}: InPageNavProps) => {
	const [activeHeadingId, setActiveHeadingId] = useState<null | string>(null);
	const [linkTree, setlinkTree] = useState<LinkTreeItem[]>([]);

	// Build the tree if links to use in the nav, from the headings found on the page
	useEffect(() => {
		const headingsContainerElement = document.querySelector(
			headingsContainerSelector
		);

		if (headingsContainerElement) {
			const headingstoExclude = headingsExcludeSelector
				? Array.prototype.slice.call(
						headingsContainerElement.querySelectorAll(headingsExcludeSelector)
				  )
				: [];

			// If we've specified a whole container to exclude,
			// search for any headings within each matched container
			if (headingsExcludeContainer) {
				const excludedElements = document.querySelectorAll(
					headingsExcludeContainer
				);
				if (excludedElements.length) {
					excludedElements.forEach((element) => {
						const headingsWithinExcludedContainer =
							element.querySelectorAll(headingsSelector);
						headingstoExclude.push(...headingsWithinExcludedContainer);
					});
				}
			}

			const headingsToUse = Array.prototype.slice
				.call(headingsContainerElement.querySelectorAll(headingsSelector))
				.filter((el) => headingstoExclude.indexOf(el) === -1);

			setlinkTree(buildLinkTree(headingsToUse));
		}

		return () => {
			setlinkTree([]);
		};
	}, [headingsContainerSelector, headingsSelector, headingsExcludeSelector]);

	// Now that we've built the tree of links, work out which is active based on scroll position
	useEffect(() => {
		const scrollHandler = throttle(() => {
			setActiveHeadingId(getActiveHeadingId(linkTree, scrollTolerance));
		}, 100);

		setActiveHeadingId(getActiveHeadingId(linkTree, scrollTolerance));
		window.addEventListener("scroll", scrollHandler, { passive: true });

		return () => {
			setActiveHeadingId(null);
			window.removeEventListener("scroll", scrollHandler);
		};
	}, [linkTree]);

	if (linkTree.length === 0) return null;

	return (
		<nav
			className={["in-page-nav", className].join(" ")}
			aria-labelledby="inpagenav-title"
			data-component="in-page-nav"
			{...rest}
		>
			<h2 id="inpagenav-title" className="in-page-nav__title">
				On this page
			</h2>
			<ol
				className="in-page-nav__list"
				aria-label="Jump to sections on this page"
			>
				{linkTree.map((link) => (
					<InPageNavLink
						key={link.href}
						link={link}
						activeHeadingId={activeHeadingId}
					/>
				))}
			</ol>
		</nav>
	);
};
