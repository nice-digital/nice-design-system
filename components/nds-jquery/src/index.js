import pluginAutoLoader from "./plugin-autoloader";
import Tabs from "./tabs.js";
import InPageNav from "./in-page-nav.js";
import Tracker from "./tracker";
import pluginizr from "./pluginizr";
import eventr from "./eventr";
import breakpoints from "../../nds-core/src/breakpoints";
import utils from "../../nds-core/src/utils";

let nice = {
	init: (el: $) => {

		// Load all component modules from this directory automatically, see http://stackoverflow.com/a/31770875/486434
		// But exclude test files http://stackoverflow.com/a/30372240
		pluginAutoLoader.load(require.context("./../components/", true, /^((?!test\.).)*\.js$/im));
		pluginAutoLoader.load(require.context("./", true, /^((?!test\.).)*\.js$/im));

		pluginAutoLoader.findPlugins(el);
	},
	// Components
	InPageNav: InPageNav,
	Tabs: Tabs,
	Tracker: Tracker,
	// Utils
	pluginAutoLoader: pluginAutoLoader,
	pluginizr: pluginizr,
	eventr: eventr,
	breakpoints,
	utils
};

$.fn.nice = function() {
	nice.init(this);
};

// Export to global namespace for precompiled usage
window.NICE = window.NICE || {};
window.NICE.nice = nice;

export {
	nice as default,
	// Components
	InPageNav,
	Tabs,
	Tracker,
	// Utils
	pluginAutoLoader,
	pluginizr,
	eventr,
	breakpoints,
	utils
};
