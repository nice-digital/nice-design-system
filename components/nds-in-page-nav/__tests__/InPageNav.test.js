import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";

import { InPageNav } from "../src/InPageNav";

describe("InPageNav", () => {
	it("should render without crashing", () => {
		const wrapper = shallow(<InPageNav />);
		expect(wrapper).toHaveLength(1);
	});

	it("should match snapshot", () => {
		const main = document.createElement("main");
		document.body.appendChild(main);

		const h2 = document.createElement("h2");
		h2.textContent = "A heading 2";
		main.appendChild(h2);

		const h3 = document.createElement("h3");
		h3.textContent = "A heading 3";
		main.appendChild(h3);

		const inPageNavContainer = document.createElement("div");
		document.body.appendChild(inPageNavContainer);

		const wrapper = mount(<InPageNav data-track={false} />, {
			attachTo: inPageNavContainer
		});
		wrapper.update();
		expect(toJson(wrapper)).toMatchSnapshot();
	});
});
