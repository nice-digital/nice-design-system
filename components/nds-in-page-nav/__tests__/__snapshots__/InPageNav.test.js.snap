// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InPageNav should match snapshot 1`] = `
<InPageNav
  data-track={false}
>
  <nav
    aria-labelledby="inpagenav-title"
    className="in-page-nav "
    data-track={false}
  >
    <h2
      className="in-page-nav__title"
      id="inpagenav-title"
    >
      On this page
    </h2>
    <ol
      aria-label="Jump to sections on this page"
      className="in-page-nav__list"
    >
      <InPageNavLink
        activeHeadingId="a-heading-3"
        key="#a-heading-2"
        link={
          Object {
            "href": "#a-heading-2",
            "subLinks": Array [
              Object {
                "href": "#a-heading-3",
                "subLinks": Array [],
                "title": "A heading 3",
              },
            ],
            "title": "A heading 2",
          }
        }
      >
        <li
          className="in-page-nav__item"
        >
          <a
            aria-current="true"
            href="#a-heading-2"
            id="inpagenav-a-heading-2"
          >
            A heading 2
          </a>
          <ol
            aria-labelledby="inpagenav-a-heading-2"
            className="in-page-nav__list"
          >
            <InPageNavLink
              activeHeadingId="a-heading-3"
              key="#a-heading-3"
              link={
                Object {
                  "href": "#a-heading-3",
                  "subLinks": Array [],
                  "title": "A heading 3",
                }
              }
            >
              <li
                className="in-page-nav__item"
              >
                <a
                  aria-current="location"
                  href="#a-heading-3"
                  id="inpagenav-a-heading-3"
                >
                  A heading 3
                </a>
              </li>
            </InPageNavLink>
          </ol>
        </li>
      </InPageNavLink>
    </ol>
  </nav>
</InPageNav>
`;
